File: server/src/main/java/com/netmarble/chat/presentation/controller/ReadStatusController.java
package com.netmarble.chat.presentation.controller;

import com.netmarble.chat.application.dto.UnreadCountResponse;
import com.netmarble.chat.application.service.ReadStatusApplicationService;
import com.netmarble.chat.presentation.exception.UnauthorizedException;
import jakarta.servlet.http.HttpSession;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * 읽음 상태 관리 REST API Controller
 */
@Slf4j
@RestController
@RequestMapping("/api/read-status")
@RequiredArgsConstructor
public class ReadStatusController {

    private final ReadStatusApplicationService readStatusApplicationService;

    private Long requireUserId(HttpSession session) {
        return Optional.ofNullable(session.getAttribute("userId"))
                .map(o -> (Long) o)
                .orElseThrow(() -> new UnauthorizedException("로그인이 필요합니다."));
    }

    /**
     * 채팅방의 메시지를 읽음 처리
     * POST /api/read-status/mark-read
     */
    @PostMapping("/mark-read")
    public ResponseEntity<Void> markAsRead(
            @RequestParam Long chatRoomId,
            HttpSession session) {
        Long userId = requireUserId(session);
        log.info("POST /api/read-status/mark-read - userId={}, chatRoomId={}", userId, chatRoomId);
        readStatusApplicationService.markAsRead(userId, chatRoomId);
        return ResponseEntity.noContent().build();
    }

    /**
     * 특정 채팅방의 읽지 않은 메시지 개수 조회
     * GET /api/read-status/unread-count?chatRoomId=
     */
    @GetMapping("/unread-count")
    public ResponseEntity<Long> getUnreadCount(
            @RequestParam Long chatRoomId,
            HttpSession session) {
        Long userId = requireUserId(session);
        log.info("GET /api/read-status/unread-count - userId={}, chatRoomId={}", userId, chatRoomId);
        long unreadCount = readStatusApplicationService.getUnreadCount(userId, chatRoomId);
        return ResponseEntity.ok(unreadCount);
    }

    /**
     * 사용자의 모든 채팅방별 읽지 않은 메시지 개수 조회
     * GET /api/read-status/unread-counts
     */
    @GetMapping("/unread-counts")
    public ResponseEntity<Map<Long, Long>> getAllUnreadCounts(HttpSession session) {
        Long userId = requireUserId(session);
        log.info("GET /api/read-status/unread-counts - userId={}", userId);
        Map<Long, Long> unreadCounts = readStatusApplicationService.getAllUnreadCounts(userId);
        return ResponseEntity.ok(unreadCounts);
    }

    /**
     * 활성 채팅방의 읽지 않은 메시지 개수 목록
     * GET /api/read-status/unread-counts-list
     */
    @GetMapping("/unread-counts-list")
    public ResponseEntity<List<UnreadCountResponse>> getUnreadCountsList(HttpSession session) {
        Long userId = requireUserId(session);
        log.info("GET /api/read-status/unread-counts-list - userId={}", userId);
        List<UnreadCountResponse> unreadCounts =
            readStatusApplicationService.getUnreadCountsForActiveChatRooms(userId);
        return ResponseEntity.ok(unreadCounts);
    }
}

File: server/chat-server/src/presentation/controller/read-status.controller.ts
import {
  Controller,
  Post,
  Get,
  Param,
  Query,
  Logger,
  HttpCode,
  HttpStatus,
  ParseIntPipe,
} from '@nestjs/common';
import { ReadStatusApplicationService } from '../../application/service/read-status-application.service.js';
import { UnreadCountResponse } from '../../application/dto/index.js';

@Controller('api/read-status')
export class ReadStatusController {
  private readonly logger = new Logger(ReadStatusController.name);

  constructor(
    private readonly readStatusService: ReadStatusApplicationService,
  ) {}

  @Post('mark-read')
  @HttpCode(HttpStatus.NO_CONTENT)
  async markAsRead(
    @Query('userId') userIdStr: string,
    @Query('chatRoomId') chatRoomIdStr: string,
  ): Promise<void> {
    await this.readStatusService.markAsRead(
      parseInt(userIdStr, 10),
      parseInt(chatRoomIdStr, 10),
    );
  }

  @Get('unread-count')
  async getUnreadCount(
    @Query('userId') userIdStr: string,
    @Query('chatRoomId') chatRoomIdStr: string,
  ): Promise<number> {
    return this.readStatusService.getUnreadCount(
      parseInt(userIdStr, 10),
      parseInt(chatRoomIdStr, 10),
    );
  }

  @Get('unread-counts/:userId')
  async getAllUnreadCounts(
    @Param('userId', ParseIntPipe) userId: number,
  ): Promise<Record<string, number>> {
    const counts = await this.readStatusService.getAllUnreadCounts(userId);
    const result: Record<string, number> = {};
    for (const [key, val] of counts) {
      result[String(key)] = val;
    }
    return result;
  }

  @Get('unread-counts-list/:userId')
  async getUnreadCountsList(
    @Param('userId', ParseIntPipe) userId: number,
  ): Promise<UnreadCountResponse[]> {
    return this.readStatusService.getUnreadCountsForActiveChatRooms(userId);
  }
}
